ansible_collections/arista/avd/plugins/plugin_utils/eos_validate_state_utils/md_report.py

```diff
@@ -0,0 +1,251 @@
+# that can be found in the LICENSE file.
+from __future__ import annotations
+
+import re
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Callable, ClassVar, Generator
+
+if TYPE_CHECKING:
+    from io import TextIOWrapper
+
+    from .results_manager import ResultsManager
+
+
+class MDReportBase(ABC):
+    """Base class for generating the validation markdown report based on test results managed by ResultsManager."""
+
+    def __init__(self, mdfile: TextIOWrapper, results: ResultsManager) -> None:
+        """Initialize the MDReportBase with an open markdown file object to write to and a ResultsManager instance.
+
+        Args:
+        ----
+            mdfile (TextIOWrapper): An open file object to write the markdown data into.
+            results (ResultsManager): The ResultsManager instance containing all test results.
+        """
+        self.mdfile = mdfile
+        self.results = results
+
+    @abstractmethod
+    def generate_section(self) -> None:
+        """Abstract method to generate a specific section of the markdown report.
+
+        Must be implemented by subclasses.
+        """
+
+    def generate_header_name(self) -> str:
+        """Generate a formatted header name based on the class name.
+
+        Returns:
+        -------
+            str: Formatted header name.
+
+        Example:
+        -------
+            `ValidateStateReport` will become Validate State Report.
+        """
+        class_name = self.__class__.__name__
+        return re.sub(r"(?<!^)(?=[A-Z])", " ", class_name).title()
+
+    def generate_table(self, table_header: list[str], row_generator: Callable[[], Generator[str, None, None]]) -> None:
+        """Generate a table with a header and multiple rows.
+
+        Args:
+        ----
+            table_header (list[str]): The table header.
+            row_generator (Callable): The rows generator function.
+        """
+        self.write_table_header(table_header)
+        for row in row_generator():
+            self.mdfile.write(row)
+        self.mdfile.write("\n")
+
+    def write_header(self, heading_level: int) -> None:
+        """Write a markdown header to the markdown file.
+
+        Args:
+        ----
+            heading_level (int): The level of the heading (1-6).
+
+        Example:
+        -------
+            ## Test Results Summary
+        """
+        # Ensure the heading level is within the valid range of 1 to 6
+        heading_level = max(1, min(heading_level, 6))
+        header_name = self.generate_header_name()
+        heading = "#" * heading_level + " " + header_name
+        self.mdfile.write(f"{heading}\n\n")
+
+    def write_table_header(self, headers: list[str]) -> None:
+        """Write a table header to the markdown file.
+
+        Args:
+        ----
+            headers (list[str]): List of header strings for the table.
+        """
+        self.mdfile.write("\n".join(headers) + "\n")
```

Comment:
is it required to be a function? It is used in one place only

Reply:
Fixed in 14811244bf088911103283bdb6dcacde39f65536